compiler/rustc_mir_build/src/builder/expr/into.rs

```diff
@@ -264,77 +290,108 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
+                    let state_ty = this.thir.exprs[state].ty;
+
+                    // the type of the value that is switched on by the `SwitchInt`
+                    let discr_ty = match state_ty {
+                        ty if ty.is_enum() => ty.discriminant_ty(this.tcx),
+                        ty if ty.is_integral() => ty,
+                        _ => todo!(),
+                    };
+
+                    let rvalue = match state_ty {
+                        ty if ty.is_enum() => Rvalue::Discriminant(state_place),
+                        ty if ty.is_integral() => Rvalue::Use(Operand::Copy(state_place)),
+                        _ => todo!(),
+                    };
+
+                    // block and arm of the wildcard pattern (if any)
+                    let mut otherwise = None;
 
                     unpack!(
                         body_block = this.in_scope(
                             (region_scope, source_info),
                             LintLevel::Inherited,
                             move |this| {
-                                let unreachable_block = this.cfg.start_new_block();
-                                this.cfg.terminate(
-                                    unreachable_block,
-                                    source_info,
-                                    TerminatorKind::Unreachable,
-                                );
+                                let mut arm_blocks = Vec::with_capacity(arms.len());
+                                for &arm in arms {
+                                    let pat = &this.thir[arm].pattern;
+                                    let deconstructed_pat = cx.lower_pat(pat);
+
+                                    match deconstructed_pat.ctor() {
+                                        Constructor::Variant(variant_index) => {
+                                            let PatKind::Variant { adt_def, .. } = pat.kind else {
+                                                unreachable!()
+                                            };
+
+                                            let discr = adt_def
+                                                .discriminant_for_variant(this.tcx, *variant_index);
+
+                                            let block = this.cfg.start_new_block();
+                                            arm_blocks.push((discr, block, arm))
+                                        }
+                                        Constructor::IntRange(int_range) => {
+                                            assert!(int_range.is_singleton());
+
+                                            let bits = state_ty.primitive_size(this.tcx).bits();
+                                            let Some(value) = int_range.lo.as_finite_int(bits)
+                                            else {
+                                                todo!()
+                                            };
```

Comment:
```suggestion
                                            let value = int_range.lo.as_finite_int(bits).unwrap();
```

this shouldn't fail for singleton ranges.